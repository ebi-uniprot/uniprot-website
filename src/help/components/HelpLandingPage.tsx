// TODO: fix import order
import { ChangeEvent, useState } from 'react';
import { DataList, SearchInput } from 'franklin-sites';

import useDataApi from '../../shared/hooks/useDataApi';

import HelpCompactResult from './shared/HelpCompactResult';

import { help as helpURL } from '../../shared/config/apiUrls';

import { HelpSearchResponse } from '../adapters/helpConverter';
import { getIdKey } from './results/Results';

const HelpLandingPage = () => {
  const [searchValue, setSearchValue] = useState<string>('');
  const dataObject = useDataApi<HelpSearchResponse>(
    searchValue && helpURL.search({ query: searchValue })
  );

  return (
    <>
      <h1>Help center</h1>
      <div>
        <SearchInput
          isLoading={dataObject.loading}
          onChange={(event: ChangeEvent<HTMLInputElement>) =>
            setSearchValue(event.target.value)
          }
          placeholder="Search"
          value={searchValue}
          autoFocus
        />

        {dataObject?.data?.results && (
          <DataList
            getIdKey={getIdKey}
            data={dataObject.data.results}
            dataRenderer={(article) => (
              <HelpCompactResult
                id={article.id}
                title={article.title}
                titleMatch={article.matches?.title?.[0]}
                contentMatch={article.matches?.content?.[0]}
              />
            )}
          />
        )}
      </div>
    </>
  );
};

export default HelpLandingPage;
