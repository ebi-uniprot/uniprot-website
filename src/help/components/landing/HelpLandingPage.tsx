import HelpAutocomplete from './HelpAutocomplete';

const HelpLandingPage = () => (
  <div className="uniprot-grid--centered">
    <h1>Help center</h1>
    <HelpAutocomplete />
  </div>
);

export default HelpLandingPage;
