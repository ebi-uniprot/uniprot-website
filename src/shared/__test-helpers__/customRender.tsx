/**
 * This tries to follow what is detailed here:
 * https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
import { render, RenderOptions } from '@testing-library/react';
import { Component, Dispatch, ReactElement, ReactNode } from 'react';
import { HelmetProvider } from 'react-helmet-async';
import {
  createMemoryHistory,
  LocationState,
  MemoryHistory,
  Route,
  Router,
} from 'react-router';
import { JsonValue, SetRequired } from 'type-fest';

import { MessagesState } from '../../messages/state/messagesInitialState';
import { MessagesAction } from '../../messages/state/messagesReducers';
import databaseInfo from '../../uniprotkb/utils/__tests__/__mocks__/databaseInfo';
import { getDatabaseInfoMaps } from '../../uniprotkb/utils/database';
import {
  MessagesDispatchContext,
  MessagesStateContext,
} from '../contexts/Messages';
import {
  DatabaseInfoMapsContext,
  UniProtDataVersionContext,
} from '../contexts/UniProtData';
import { JobAction } from '../workers/jobs/state/actionHandler';
import { JobsState } from '../workers/jobs/state/jobsInitialState';

type ExtraRenderOptions = {
  // For react-router
  history?: MemoryHistory<LocationState>;
  path?: string;
  // For custom user preferences (used by useLocalStorage)
  initialLocalStorage?: Record<string, JsonValue>;
  // For app context
  messagesState?: MessagesState;
  messagesDispatch?: jest.Mock<Dispatch<MessagesAction>>;
  JobsState?: JobsState;
  toolsDispatch?: jest.Mock<Dispatch<JobAction>>;
};

type WrapperProps = { children: ReactNode } & RenderOptions &
  SetRequired<
    ExtraRenderOptions,
    | 'history'
    | 'initialLocalStorage'
    | 'messagesState'
    | 'messagesDispatch'
    | 'JobsState'
    | 'toolsDispatch'
  >;

const dbInfoMaps = getDatabaseInfoMaps(databaseInfo);
const uniProtDataVersion = {
  releaseNumber: '2023_01',
  releaseDate: new Date('01-March-2023'),
};

class Wrapper extends Component<WrapperProps> {
  constructor(props: WrapperProps) {
    super(props);
    for (const [key, value] of Object.entries(props.initialLocalStorage)) {
      window.localStorage.setItem(key, JSON.stringify(value));
    }
  }

  // eslint-disable-next-line class-methods-use-this
  componentWillUnmount() {
    window.localStorage.clear();
  }

  render() {
    const { children, path, history, messagesDispatch, messagesState } =
      this.props;
    return (
      <HelmetProvider>
        <Router history={history}>
          <MessagesDispatchContext.Provider value={messagesDispatch}>
            <MessagesStateContext.Provider value={messagesState}>
              <DatabaseInfoMapsContext.Provider value={dbInfoMaps}>
                <UniProtDataVersionContext.Provider value={uniProtDataVersion}>
                  {path ? (
                    <Route path={path} render={() => children} />
                  ) : (
                    children
                  )}
                </UniProtDataVersionContext.Provider>
              </DatabaseInfoMapsContext.Provider>
            </MessagesStateContext.Provider>
          </MessagesDispatchContext.Provider>
        </Router>
      </HelmetProvider>
    );
  }
}

const customRender = (
  ui: ReactElement,
  {
    route = '',
    path,
    history = createMemoryHistory({ initialEntries: [route] }),
    initialLocalStorage = {},
    messagesState = {},
    messagesDispatch = jest.fn(),
    JobsState = {},
    toolsDispatch = jest.fn(),
    ...options
  }: RenderOptions &
    ExtraRenderOptions & {
      route?: string;
    } = {}
) => ({
  ...render(ui, {
    wrapper: (props) => (
      <Wrapper
        path={path}
        history={history}
        initialLocalStorage={initialLocalStorage}
        messagesState={messagesState}
        messagesDispatch={messagesDispatch}
        JobsState={JobsState}
        toolsDispatch={toolsDispatch}
        {...props}
      />
    ),
    ...options,
  }),
  history,
  messagesDispatch,
  toolsDispatch,
});

export default customRender;
