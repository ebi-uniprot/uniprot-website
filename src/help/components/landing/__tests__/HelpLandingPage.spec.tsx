import customRender from '../../../../shared/__test-helpers__/customRender';

import HelpLandingPage from '../HelpLandingPage';

describe('HelpLandingPage', () => {
  it('should render', () => {
    const { asFragment } = customRender(<HelpLandingPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});
