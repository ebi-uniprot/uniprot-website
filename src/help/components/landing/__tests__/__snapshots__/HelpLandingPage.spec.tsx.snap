// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpLandingPage should render 1`] = `
<DocumentFragment>
  <main
    class="uniprot-grid uniprot-grid--centered help-landing"
  >
    <div
      class="uniprot-grid-cell--span-12 page-intro"
    >
      <h1
        class=""
      >
        Help center
      </h1>
    </div>
    <div
      class="uniprot-grid-cell--span-9"
    >
      <div
        class="help-quick-search"
      >
        <span
          class="search-input"
        >
          <input
            data-testid="search-input"
            placeholder="Search"
            type="search"
            value=""
          />
          <span
            class="search-input__suffix"
            data-testid="search-input-suffix"
            role="presentation"
          >
            <test-file-stub
              height="14"
              width="14"
            />
          </span>
        </span>
      </div>
      <div
        class="help-landing__tiles"
      >
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            UniProt data
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <img
              alt=""
              height="62"
              src="test-file-stub"
              width="52"
            />
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="/help/uniprotkb_manual"
              >
                UniProtKB
              </a>
            </li>
            <li>
              <a
                href="/help/proteomes_manual"
              >
                Proteomes
              </a>
            </li>
            <li>
              <a
                href="/help/uniref"
              >
                UniRef
              </a>
            </li>
            <li>
              <a
                href="/help/uniparc"
              >
                UniParc
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help?query=data"
          >
            More
          </a>
        </div>
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            Get started
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <img
              alt=""
              height="52"
              src="test-file-stub"
              width="52"
            />
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="/help/find_your_protein"
              >
                Find your protein
              </a>
            </li>
            <li>
              <a
                href="/help/explore_uniprotkb_entry "
              >
                Explore UniProtKB entry
              </a>
            </li>
            <li>
              <a
                href="/help/explore_proteomes "
              >
                Explore proteomes
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help?query=*&facets=category:Get started"
          >
            More
          </a>
        </div>
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            Training
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <img
              alt=""
              height="40"
              src="test-file-stub"
              width="52"
            />
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="https://www.ebi.ac.uk/training/online/courses/uniprot-quick-tour"
              >
                Online training
              </a>
            </li>
            <li>
              <a
                href="https://www.ebi.ac.uk/training/search-results?query=uniprot&domain=ebiweb_training&page=1&facets="
              >
                Courses
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help?query=training"
          >
            More
          </a>
        </div>
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            Technical corner
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <img
              alt=""
              height="41"
              src="test-file-stub"
              width="52"
            />
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="/help/programmatic_access"
              >
                Programmatic access
              </a>
            </li>
            <li>
              <a
                href="/help/programmatic_access"
              >
                UniProt JAPI
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help/technical"
          >
            More
          </a>
        </div>
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            Biocuration
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <img
              alt=""
              height="59"
              src="test-file-stub"
              width="52"
            />
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="/help/biocuration"
              >
                UniProt biocuration
              </a>
            </li>
            <li>
              <a
                href="/help?query=*&facets=category:biocuration"
              >
                Biocuration projects
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help?query=*&facets=category:biocuration"
          >
            More
          </a>
        </div>
        <div
          class="illustrated-list-tile"
        >
          <h2
            class="illustrated-list-tile__heading medium"
          >
            About us
          </h2>
          <div
            class="illustrated-list-tile__image"
          >
            <div
              style="opacity: 0.8; color: rgb(40, 170, 80);"
            >
              <test-file-stub />
            </div>
          </div>
          <ul
            class="illustrated-list-tile__list no-bullet"
          >
            <li>
              <a
                href="/help/about"
              >
                About UniProt
              </a>
            </li>
            <li>
              <a
                href="/help/publications"
              >
                How to cite us
              </a>
            </li>
          </ul>
          <a
            class="illustrated-list-tile__more"
            href="/help?query=*&facets=category:About UniProt"
          >
            More
          </a>
        </div>
      </div>
    </div>
    <div
      class="uniprot-grid-cell--span-3"
    >
      <section
        class="help-landing__faqs"
      >
        <h2
          class="small"
        >
          Suggested FAQs
        </h2>
        <ul
          class="no-bullet"
        >
          <li>
            <a
              href="/help/disease_query"
            >
              How can I get all the proteins involved in a given disease?
            </a>
          </li>
          <li>
            <a
              href="/help/canonical_and_isoforms"
            >
              What is the canonical sequence?
            </a>
          </li>
          <li>
            <a
              href="/help/reference_proteome"
            >
              What are reference proteomes?
            </a>
          </li>
        </ul>
        <a
          href="/help?query=*&facets=category:faq"
        >
          View all FAQs
        </a>
      </section>
      <section
        class="help-landing__contact"
      >
        <div>
          <p>
            <a
              href="/contact"
            >
              Get in touch 
              <test-file-stub
                width="2ch"
              />
            </a>
          </p>
          <p
            class="social"
          >
            <a
              class="external-link"
              href="https://twitter.com/uniprot"
              rel="noopener"
              target="_blank"
              title="UniProt posts on X (formerly Twitter)"
            >
              <test-file-stub
                width="3ch"
              />
            </a>
            <a
              class="external-link"
              href="https://www.facebook.com/uniprot.org"
              rel="noopener"
              target="_blank"
              title="UniProt posts on Facebook"
            >
              <test-file-stub
                width="3ch"
              />
            </a>
            <a
              class="external-link"
              href="https://www.youtube.com/user/uniprotvideos"
              rel="noopener"
              target="_blank"
              title="UniProt videos on YouTube"
            >
              <test-file-stub
                width="3ch"
              />
            </a>
            <a
              class="external-link"
              href="https://insideuniprot.blogspot.com/"
              rel="noopener"
              target="_blank"
              title="UniProt blog"
            >
              <test-file-stub
                width="3ch"
              />
            </a>
            <a
              class="external-link"
              href="https://groups.google.com/forum/#!forum/ebi-proteins-api"
              rel="noopener"
              target="_blank"
              title="UniProt Google Group"
            >
              <test-file-stub
                width="3ch"
              />
            </a>
          </p>
        </div>
      </section>
      <section
        class="help-landing__videos"
      >
        <h4>
          Help videos
        </h4>
        <div
          class="help-landing__videos__list"
        >
          <lite-youtube
            playlabel="How to search UniProtKB"
            style="background-image: url(https://i.ytimg.com/vi/yp1O1gDK8oA/hqdefault.jpg);"
            videoid="yp1O1gDK8oA"
          >
            <button
              class="lty-playbtn"
              type="button"
            >
              <span
                class="visually-hidden"
              >
                How to search UniProtKB
              </span>
            </button>
          </lite-youtube>
        </div>
      </section>
    </div>
  </main>
</DocumentFragment>
`;
